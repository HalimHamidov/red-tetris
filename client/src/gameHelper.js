import { Tetrominos } from "./tetrominos";
export const S_WIDTH = 10;
export const S_HEIGHT = 20;


export const Createstage = (h = S_HEIGHT, w = S_WIDTH) =>
  Array.from(Array(h), () => new Array(w).fill([0, "clear"]));

export const checkcollision = (player, stage, { x: moveX, y: moveY }) => {
  for (let y = 0; y < player.tetromino.length; y++) {
    for (let x = 0; x < player.tetromino.length; x++) {
      if (player.tetromino[y][x] !== 0) {
        if (
          !stage[y + player.pos.y + moveY] ||
          !stage[y + player.pos.y + moveY][x + player.pos.x + moveX] ||
          stage[y + player.pos.y + moveY][x + player.pos.x + moveX][1] !==
            "clear"
        )
          return true;
      }
    }
  }
  return false;
};
