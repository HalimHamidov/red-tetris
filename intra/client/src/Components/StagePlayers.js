import React from "react";
import { StyledSpecter } from "./styling/StyledStage";

const StagePlayers = ({ stage, user, score }) => {
  if (!stage) {
    return <span>Loading...</span>;
  }
  return (
    <StyledSpecter>
      <div className="flex align-center justify-sa">
        <h3 style={{ color: "#fff" }}>{user}</h3>
        <h2 style={{ color: "#fff" }}>{score}</h2>
      </div>
      <table style={{ border: "1px solid" }}>
        {stage.map((row, i) => {
          return (
            <tbody key={`s-${i}`}>
              <tr style={{ height: "15px", padding: "0px", margin: "0px" }}>
                {row.map((cell, x) =>
                  cell[0] === 0 ? (
                    <td
                      data-testid={`stagePColE-${i}`} 
                      style={{
                        border: 0,
                        width: "15px",
                        padding: "0px",
                        margin: "0px",
                        height: "2px",
                        backgroundColor: "#000",
                        opacity: "30%",
                      }}
                      key={`col-${x}`}
                      type={cell[0]}
                    ></td>
                  ) : (
                    <td
                      data-testid={`stagePColF-${i}`}
                      style={{
                        border: 0,
                        width: "27px",
                        padding: "8px",
                        margin: "0px",
                        height: "2px",
                        backgroundColor: "#b3ffff",
                      }}
                      key={x}
                      type={cell[0]}
                    ></td>
                  )
                )}
              </tr>
            </tbody>
          );
        })}
      </table>
    </StyledSpecter>
  );
};

export default StagePlayers;
