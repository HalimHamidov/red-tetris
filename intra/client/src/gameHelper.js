import { Tetrominos } from "./tetrominos";
export const S_WIDTH = 10;
export const S_HEIGHT = 20;


export const Createstage = (h = S_HEIGHT, w = S_WIDTH) =>
    Array.from(Array(h), () => new Array(w).fill([0, "clear"]));

export const checkcollision = (player, stage, { x: moveX, y: moveY }) => {
    // Использование цикла for для возможности возврата (и прерывания). Невозможно с forEach
    for (let y = 0; y < player.tetromino.length; y++) {
        for (let x = 0; x < player.tetromino.length; x++) {
            // 1. Проверяем, что мы находимся в текущей ячейке Тетромино
            if (player.tetromino[y][x] !== 0) {
                // 2. Проверяем, что наш ход находится внутри высоты игровых площадок (y)
                // Что мы не проходим низ игровой зоны
                if (!stage[y + player.pos.y + moveY] ||
                    // 3. Проверяем, что наш ход находится внутри ширины игровых зон (x)
                    !stage[y + player.pos.y + moveY][x + player.pos.x + moveX] ||
                    // 4. Проверяем, что ячейка, в которую мы двигаемся, не установлена ​​на очистку(clear)
                    stage[y + player.pos.y + moveY][x + player.pos.x + moveX][1] !== "clear"
                )
                    return true;
            }
        }
    }
    // 5. Если все вышеперечисленное неверно
    return false;
};