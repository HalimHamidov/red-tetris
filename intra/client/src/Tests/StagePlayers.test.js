import renderer from "react-test-renderer";
import StagePlayers from "../Components/StagePlayers";
import { render, waitFor } from "@testing-library/react";
import { Createstage } from "../gameHelper";

test("StagePlayers render test ", () => {
    const tree = renderer.create( < StagePlayers / > ).toJSON();
    expect(tree).toMatchSnapshot();
});
test("StagePlayers loading test", async() => {
            const { findByText } = render( < StagePlayers stage = { null }
                />);
                await waitFor(() => {
                    expect(findByText("Loading...")).toBeDefined();
                });
            }); test("StagePlayers stage is set test username", async() => {
            const testUser = "testUser";
            const testScore = "800";
            const { findByText } = render( <
                StagePlayers stage = { Createstage() }
                user = { testUser }
                />
            );
            await waitFor(() => {
                expect(findByText("testUser")).toBeDefined();
            });
        }); test("StagePlayers stage is set test specters", async() => {
                const { getByTestId } = render( < StagePlayers stage = {
                        [
                            [
                                [0, "clear"]
                            ]
                        ] }
                    />);
                    await waitFor(() => {
                        expect(getByTestId("stagePColE-0")).toBeTruthy();
                    });
                }); test("StagePlayers stage is set test specters piece collided", async() => {
                    const { getByTestId } = render( < StagePlayers stage = {
                            [
                                [
                                    ["T", "clear"]
                                ]
                            ] }
                        />);
                        await waitFor(() => {
                            expect(getByTestId("stagePColF-0")).toBeTruthy();
                        });
                    });