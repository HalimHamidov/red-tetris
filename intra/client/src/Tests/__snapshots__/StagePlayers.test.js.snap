// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StagePlayers render test  1`] = `
<span>
  Loading...
</span>
`;
